tedana/utils.py

```diff
@@ -49,39 +49,73 @@ def reshape_niimg(data):
-        Boolean array of voxels that have sufficient signal in at least one
-        echo
+        Boolean array of voxels that have sufficient signal in at least ``threshold`` echos.
     masksum : (S,) :obj:`numpy.ndarray`
-        Valued array indicating the number of echos with sufficient signal in a
-        given voxel. Only returned if `getsum = True`
+        Valued array indicating the number of echos with sufficient signal in a given voxel.
+
+    Notes
+    -----
+    The adaptive mask is constructed from the following method:
+
+    1.  Count the total number of echoes in each voxel that have "good" data.
+        This method assumes that an exemplar voxel's signal at later echoes is also reasonable,
+        and that any voxels whose values at a given echo are less than 1/3 of the exemplar voxel's
+        values at that echo are affected by dropout.
+
+        This method uses distributions of values across the mask.
+        Therefore, it is sensitive to the quality of the mask;
+        a bad mask may result in a bad adaptive mask.
+
+        This method is implemented as follows:
+
+        a.  Calculate the 33rd percentile of values in the first echo,
+            based on the voxel-wise mean over time.
+
+            -   The 33rd percentile is arbitrary.
+            -   The percentile is calculated only across voxels with non-zero values.
+                However, it is rare for voxels to have values of zero in the first echo,
+                so this exclusion will likely not have a major effect.
+        b.  Identify the voxel where the first echo's mean value is equal to the 33rd percentile.
+            Basically, this identifies an "exemplar" voxel reflecting the 33rd percentile.
+
+            -   If more than one voxel has a value exactly equal to the 33rd percentile,
+                keep all of them.
+        c.  Calculate 1/3 of the mean value of the exemplar voxel for each echo.
+
+            -   This is the threshold for "good" data.
+            -   The 1/3 value is arbitrary.
+            -   If there was more than one exemplar voxel,
+                retain the the highest value across the exemplars for each echo.
+        d.  For each voxel, count the number of echoes that have a mean value greater than the
+            corresponding echo's threshold.
     """
     RepLGR.info(
-        "An adaptive mask was then generated, in which each voxel's "
-        "value reflects the number of echoes with 'good' data."
+        "An adaptive mask was then generated, "
+        "in which each voxel's value reflects the number of echoes with 'good' data."
     )
+    mask = reshape_niimg(mask).astype(bool)
+    data = data[mask, :, :]
 
-    # take temporal mean of echos and extract non-zero values in first echo
+    # take temporal mean of echos and extract non-zero voxels in first echo
     echo_means = data.mean(axis=-1)  # temporal mean of echos
     first_echo = echo_means[echo_means[:, 0] != 0, 0]
```

Comment:
Why do we limit to non-zero voxels here?

Reply:
A truly zero voxel is one that was already masked out before this step so this is removing voxels with no data. I suspect this bug first arose because, at some point, the data were already masked before this step was run so that only voxels within the mask (i.e. in the brain) were used.

Reply:
Thanks! I've expanded the comment a bit.